trieste/space.py

```diff
@@ -329,3 +346,124 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class ProductSearchSpace(SearchSpace):
+    r"""
+    Product :class:`SearchSpace` consisting of a product of
+    multiple :class:`SearchSpace`s. This class provides functionality for
+    accessing either the resulting combined search space or each individual space.
+
+    Note that this class assumes that individual points in product spaces are
+    represented with their inputs in the same order as specified when initializing
+    the space.
+    """
+
+    def __init__(self, spaces: list[tuple[str, SearchSpace]]):
+        r"""
+        :param spaces: A list of tag-space tuples summarizing the names and domains of
+            the space's subspaces.
+        """
+
+        tags = [tag for (tag, space) in spaces]
+        tf.debugging.assert_equal(
+            len(tags),
+            len(set(tags)),
+            message=f"subspace names must be unique, received {tags}.",
+        )
+
+        self._tags = tags
+        self._spaces = dict(zip(self._tags, [space for (tag, space) in spaces]))
```

Comment:
You can initialise a dict directly from a list of tupes.

```suggestion
        self._spaces = dict(spaces)
```